app/src/main/java/androidx/preference/PrivateKeyPickerPreference.kt

```diff
@@ -0,0 +1,119 @@
+package androidx.preference
+
+import android.content.Context
+import android.util.AttributeSet
+import androidx.core.content.edit
+import com.ubergeek42.WeechatAndroid.R
+import com.ubergeek42.WeechatAndroid.utils.AndroidKeyStoreUtils
+import com.ubergeek42.WeechatAndroid.utils.AndroidKeyStoreUtils.InsideSecurityHardware
+import com.ubergeek42.WeechatAndroid.utils.Constants
+import com.ubergeek42.WeechatAndroid.utils.Utils
+import com.ubergeek42.WeechatAndroid.utils.edDsaKeyPairToSshLibEd25519KeyPair
+import com.ubergeek42.WeechatAndroid.utils.makeKeyPair
+import com.ubergeek42.WeechatAndroid.utils.toReader
+import com.ubergeek42.cats.Kitty
+import com.ubergeek42.cats.Root
+import com.ubergeek42.weechat.relay.connection.SSHConnection
+import org.bouncycastle.jcajce.interfaces.EdDSAKey
+import java.io.IOException
+import java.nio.charset.StandardCharsets
+import java.security.GeneralSecurityException
+import java.security.KeyPair
+
+
+// Regarding the accidental override suppressing annotation, see:
+//   https://youtrack.jetbrains.com/issue/KT-12993
+//   https://youtrack.jetbrains.com/issue/KT-46621
+@Suppress("ACCIDENTAL_OVERRIDE")
+class PrivateKeyPickerPreference(context: Context?, attrs: AttributeSet?) :
+        PasswordedFilePickerPreference(context, attrs) {
+
+    // First, try decoding the key pair using our SSH library.
+    // If that fails due to unsupported key formats, try decoding the key pair using BC.
+    //
+    // Then, check if we have a pair of EdDSAKey. It is a valid key pair,
+    // but since sshlib is not friendly with BC, it can't use it.
+    // Assume it's a pair of Ed25519 keys, and covert it to sshlib format.
+    //
+    // Finally, try putting the key inside security hardware, serializing it if that fails.
```

Comment:
"secure hardware"

I've noticed this typo is already throughout existing code, even though Android API confirms the right phrase is *secure* hardware. Either fix it everywhere in this commit, or ignore this comment.